Fix #11: Add source ref

anilCSE opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes: #11 

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmwasm/wasmd/blob/master/CONTRIBUTING.md#pr-targeting))

- [x] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

ethanfrey commented on x/wasm/client/cli/tx.go:
```diff
@@ -21,13 +22,20 @@ import (
 const maxSize = 400 * 1024
 
+// whitelist
+var validBuildTags = map[string]bool{
```
Nice check, but I would prefer to whitelist all in valid format on the server, and let the clients evolve faster (rather than need a blockchain upgrade for 0.6.1 or 0.6.2

ethanfrey commented on x/wasm/client/cli/tx.go:
```diff
@@ -62,6 +70,26 @@ func StoreCodeCmd(cdc *codec.Codec) *cobra.Command {
+
+			// ensure source to be a valid uri
+			if source != "" {
+				_, err := url.Parse(source)
```
Yes

ethanfrey commented on x/wasm/client/cli/tx.go:
```diff
@@ -62,6 +70,26 @@ func StoreCodeCmd(cdc *codec.Codec) *cobra.Command {
+			// ensure builder to be a valid build tag
+			if builder != "" {
+				if !validBuildTags[builder] {
+					return fmt.Errorf("invalid tag supplied for builder %s", source)
```
As mentioned above, cab you use a regexp here instead of whitelist?

ethanfrey commented on x/wasm/internal/keeper/genesis.go:
```diff
@@ -14,7 +14,7 @@ import (
 func InitGenesis(ctx sdk.Context, keeper Keeper, data types.GenesisState) {
 	for _, code := range data.Codes {
-		newId, err := keeper.Create(ctx, code.CodeInfo.Creator, code.CodesBytes)
+		newId, err := keeper.Create(ctx, code.CodeInfo.Creator, code.CodesBytes, code.CodeInfo.Source, code.CodeInfo.Builder)
```
Can you update the genesis test to include these fields?

anilCSE replied:
Done

ethanfrey commented on x/wasm/internal/keeper/keeper_test.go:
```diff
@@ -138,7 +138,7 @@ func TestExecute(t *testing.T) {
 	require.NoError(t, err)
 
-	contractID, err := keeper.Create(ctx, creator, wasmCode)
+	contractID, err := keeper.Create(ctx, creator, wasmCode, "", "")
```
How about setting some of these to non empty in the tests

anilCSE commented:
> Nice start, this is what I had in mind.
> 
> However, as I see it, the format is only enforced in the cli and not in the keeper. (Unkess I missed that). Can you enforce that in eg. MsgCreate.ValidateBasic() and add some tests for that?
> 
> That an a few more tests and this is ready to merge

Updated `msg.ValidateBasic()` to handle this at `cli` and `handler` level. Also moved build regex from `cli` to `msg.ValidateBasic()`. May be we can replace with better regex expression.

ethanfrey commented on x/wasm/client/cli/tx.go:
```diff
@@ -62,11 +61,9 @@ func StoreCodeCmd(cdc *codec.Codec) *cobra.Command {
 				return err
 			}
 
-			// limit the input size
```
Why is this check removed?

Otherwise, the function looks good

ethanfrey replied:
Nevermind... I saw this is now handled by `ValidateBasic`. 
Very proper to remove it here to avoid duplication.

ethanfrey commented on x/wasm/handler.go:
```diff
@@ -40,7 +40,12 @@ func NewHandler(k Keeper) sdk.Handler {
 
 func handleStoreCode(ctx sdk.Context, k Keeper, msg *MsgStoreCode) sdk.Result {
-	codeID, err := k.Create(ctx, msg.Sender, msg.WASMByteCode)
+	_err := msg.ValidateBasic()
```
Why `_err`?
Custom is to use `sdkerr` for this case? (As `err` has different type, so you need two)

ethanfrey commented on x/wasm/internal/types/msg.go:
```diff
@@ -26,9 +36,26 @@ func (msg MsgStoreCode) ValidateBasic() sdk.Error {
+	if msg.Source != "" {
+		fmt.Println("source: ", msg.Source)
+
+		if !govalidator.IsURL(msg.Source) {
```
I'm not strongly against adding more dependencies, but I like to see the utility.
Bringing in govalidator for one line seems too much.

You can use: https://golang.org/pkg/net/url/#Parse from stdlib, which should have the same functionality

anilCSE replied:
We used `net/url.Parse()` previously but the method has a few weaknesses. If you change the url value to `somerandomstring#43243@` also `net/url.Parse()` function will still pass the broken URL as `valid`. This is NOT the right way to validate URL.

anilCSE replied:
May be we can do the following.
- check if url is reachable
- check if the url is absolute (we shouldn't encourage relative urls for source anyway)

anilCSE replied:
Added these changes and updated genesis tests to cover these cases.

ethanfrey commented on x/wasm/internal/types/msg.go:
```diff
@@ -1,17 +1,27 @@
 const (
-	MaxWasmSize = 500 * 1024
+	MaxWasmSize   = 500 * 1024
+	BuildTagRegex = "cosmwasm-opt:"
```
Maybe `"^cosmwasm-opt:"` so it can't start with other characters? Otherwise, that seems fine enough... some limits but not terribly restrictive
